operator/cmd/mesh/manifest-validate.go

```diff
@@ -0,0 +1,345 @@
+// Copyright 2020 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mesh
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gogo/protobuf/types"
+
+	"istio.io/istio/operator/pkg/util"
+	"istio.io/istio/operator/pkg/validate"
+	"istio.io/istio/pkg/config/validation"
+)
+
+const (
+	traceSamplingMin float64 = 0.0
+	traceSamplingMax float64 = 100.0
+)
+
+var (
+	// Boolean values for --set flags
+	boolValues = []bool{true, false}
```

Comment:
I am a bit confused by this and the meshconfig ones like outboundTrafficPolicyMode below. My understanding (which may be wrong), is we already have the value_types.proto that is typed. So we have some field like `foo bool` in the proto, and it will handle the validation for us. The same applies to enums like `outboundTrafficPolicyMode`.

For things like imagePullPolicy this is great, but I am not sure if we need the bool/enum ones? Is the existing validation not working/sufficient?

Reply:
And for the cases where we do have things like imagePullPolicy, I wonder if we should be using something like https://github.com/envoyproxy/protoc-gen-validate to have the validation inline with the rest of the spec?

Reply:
> we already have the value_types.proto that is typed. So we have some field like `foo bool` in the proto, and it will handle the validation for us. 
> 
> For things like imagePullPolicy this is great, but I am not sure if we need the bool/enum ones? Is the existing validation not working/sufficient?

The existing validation for bool lacks context and does not properly point which field has an invalid value. There is already a discussion that happened in this #22286 issue.

If you pass an invalid value by mistake for bool fields along with other valid fields. The existing validation throws `can not unmarshal string into Go value of type bool` which is good but does not provide which field has invalid values and what are the possible values we can pass.

The whole idea of static validation is to provide the following.

1. Catch error as early as possible.
2. Show which flag has invalid values.
3. Display possible values for the specific flag.

I agree we can use existing validation for `enum` fields. I tried validation with and without static checker and here is the output for both.

Existing validation for enum field
```go
$ istioctl manifest apply --set values.global.outboundTrafficPolicy.mode=ALLOW

Error: failed to apply manifests: validation errors (use --force to override): 
unknown value "ALLOW" for enum v1alpha1.OutboundTrafficPolicyConfig_Mode
```

Static validation with proper context and possible options that user can pass.

```go
$ istioctl manifest apply --set values.global.outboundTrafficPolicy.mode=ALLOW

Error: failed to apply manifests: 
 Unsupported value: "ALLOW", supported values for: "values.global.outboundTrafficPolicy.mode" is "REGISTRY_ONLY, ALLOW_ANY"
```
We can choose between which one looks appropriate for user to quickly resolve the issue.

Reply:
> And for the cases where we do have things like imagePullPolicy, I wonder if we should be using something like https://github.com/envoyproxy/protoc-gen-validate to have the validation inline with the rest of the spec?

I will take a look and see if we can utilize it. Thanks

Reply:
Thanks for the explanation. I think its clear what you have done here gives a better UX. I would love to see us have some way to have a single source of truth though if possible, since right now we will have to duplicate the information in the proto and here, and we will likely not have full coverage.

Thats the ideal case though, I have no idea how feasible it is